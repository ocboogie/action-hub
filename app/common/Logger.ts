type IListener = (msg: string) => void;

interface ILoggedError {
  category: string;
  msg: string;
}

export default class Logger {
  public history: ILoggedError[] = [];
  public historyLimit = 500;

  private categories: Map<string, IListener[]> = new Map();

  constructor(categories: string[]) {
    categories.forEach(category => {
      this.categories.set(category, []);
    });
  }

  public hasCategory(category: string): boolean {
    return this.categories.has(category);
  }

  public addListener(category: string, listener: IListener): () => void {
    const foundCategory = this.findCategory(category);
    foundCategory.push(listener);
    return () => {
      const index = foundCategory.indexOf(listener);
      if (index > -1) {
        foundCategory.splice(index, 1);
      }
    };
  }

  public report(category: string, msg: string) {
    this.history.unshift({ category, msg });
    if (this.history.length > this.historyLimit) {
      this.history = this.history.slice(0, this.historyLimit);
    }

    const foundCategory = this.findCategory(category);
    foundCategory.forEach(listener => {
      listener(msg);
    });
  }

  private findCategory(category: string): IListener[] {
    const foundCategory = this.categories.get(category);
    if (!foundCategory) {
      throw new TypeError(`No category named "${category}"`);
    }
    return foundCategory;
  }
}
