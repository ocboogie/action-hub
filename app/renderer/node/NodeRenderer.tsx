import * as joi from "joi";
import * as React from "react";

import Logger from "../../common/Logger";
import { INode } from "./index";

export interface INodeRendererProps {
  node: INode;
  args?: object;
  logger?: Logger;
}

export default class NodeRenderer extends React.Component<INodeRendererProps> {
  public render() {
    const logger = this.props.logger || window.logger;
    const node = this.props.node;
    const NodeComponent = node.component;

    const error = node.argSchema.validate(this.props.args).error;
    if (error) {
      if (!logger) {
        throw error;
      }
      logger.report(
        "error",
        "Incorrect arguments passed to node",
        error.message
      );
      return <div />;
    }

    return (
      <div>
        <NodeComponent {...this.props.args} />
      </div>
    );
  }
}
