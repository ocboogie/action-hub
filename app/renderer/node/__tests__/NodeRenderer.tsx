// tslint:disable-next-line:no-implicit-dependencies
import { mount, shallow } from "enzyme";
import * as React from "react";

import Logger from "../../../common/Logger";
import NodeRenderer from "../NodeRenderer";
import mockNode from "./resources/mockNode";

const logger = new Logger(["error"]);

test("renders node", () => {
  const wrapper = mount(
    <NodeRenderer node={mockNode} args={{ text: "foo" }} />
  );

  expect(wrapper.contains(<h1>foo</h1>)).toBeTruthy();
});

test("throws when arguments are invalid", () => {
  expect(() => {
    const wrapper = shallow(
      <NodeRenderer node={mockNode} args={{ foo: "bar" }} />
    );
  }).toThrowErrorMatchingSnapshot();
});

test("logs when arguments are invalid", () => {
  const mockListener = jest.fn();
  logger.addListener("error", mockListener);

  const wrapper = shallow(
    <NodeRenderer logger={logger} node={mockNode} args={{ foo: "bar" }} />
  );

  expect(mockListener.mock.calls[0][0]).toMatchObject({
    title: "Incorrect arguments passed to node"
  });
});
