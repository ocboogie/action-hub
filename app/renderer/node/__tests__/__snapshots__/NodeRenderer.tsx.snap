// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws when arguments are invalid 1`] = `"child \\"text\\" fails because [\\"text\\" is required]"`;
